src/PageHeader/PageHeader.tsx

```diff
@@ -56,19 +57,90 @@ export type PageHeaderProps = {
   as?: React.ElementType | 'header' | 'div'
 } & SxProp
 
-const Root: React.FC<React.PropsWithChildren<PageHeaderProps>> = ({children, sx = {}, as = 'div'}) => {
-  const rootStyles = {
-    display: 'flex',
-    flexDirection: 'column',
-    // TODO: We used hard-coded values for the spacing and font size in this component. Update them to use new design tokens when they are ready to use.
-    gap: '0.5rem',
-  }
-  return (
-    <Box as={as} sx={merge<BetterSystemStyleObject>(rootStyles, sx)}>
-      {children}
-    </Box>
-  )
-}
+const Root = React.forwardRef<HTMLDivElement, React.PropsWithChildren<PageHeaderProps>>(
+  ({children, sx = {}, as = 'div'}, forwardedRef) => {
+    const rootRef = useProvidedRefOrCreate<HTMLDivElement>(forwardedRef as React.RefObject<HTMLDivElement>)
+
+    const [contextArea, setContextArea] = React.useState<string>('')
+    const [titleArea, setTitleArea] = React.useState<string>('leadingAction titleArea trailingAction actions')
+    const [description, setDescription] = React.useState<string>('')
+    const [navigation, setNavigation] = React.useState<string>('')
+
+    const gridTemplateAreas = `${contextArea !== '' ? `'${contextArea}' ` : ''}${titleArea ? `'${titleArea}' ` : ''}${
+      description ? `'${description}' ` : ''
+    }${navigation ? `'${navigation}' ` : ''}`.trim()
+
+    const rootStyles = {
+      display: 'grid',
+      // We have 4 columns.
+      gridTemplateColumns: 'auto auto auto 1fr',
+      // We are rendering the template areas dynamically based on the children's visibility.
+      gridTemplateAreas,
+      gap: 'var(--stack-gap-condensed, 0.5rem)',
+    }
+
+    const [visibility, setVisibility] = React.useState({
```

Comment:
This needs to be based on the components' visibility rather than checking if the component exists. Because component can exist but not be visible in a given viewport. (I.e. ContextArea can exist in the code and renders in the DOM in all viewport sizes but only visible in narrow viewports.)

Reply:
I think we should try and come up with a way to do this using only CSS. I'd hate to introduce a bunch of re-renders caused by state changes.

Reply:
I know 😭 I wasn't able to think of a way to do it with CSS but let me re-think of any new ideas today. If not, we can work on it tomorrow in our pairing session 🙌🏻